### ISSUE TITLE
2.x Roadmap/Todos

### ISSUE THREAD

**Issue by tbreuss** [getherbie/herbie#82]
Todo list for the upcoming 2.x release.

**Nice-to-have**
- [x] Implement PSR-14 Event Dispatcher <https://www.php-fig.org/psr/psr-14/>
- [x] Add aliases to system info twig function
- [x] Rename `example` folder to `website`
- [x] Rename `src` folder to `system`
- [x] Rename `sysplugins` folder to `plugins`
- [x] LowerCamelCase member vars in Page class. This is already done, since all of this is possible: page.content_type, page.contentType, page.getContentType()
- [x] Consistent naming for methods (applications instead of app, eventFilters for events, etc.)
- [x] Integrate GitHub atom feed into the homepage of the website
- [x] Implement a query builder class for filtering, sorting und limiting of iterable data structures
- [x] Add readme to all system plugins (https://github.com/getherbie/herbie/tree/2.x/plugins)

**Suggestions**
- [x]  Group twig filters, functions and tests logically to make it more clear for the user (i.e. nav_sitemap, nav_breadcrumb, etc.)
- [x] Replace InterceptingFilters with PSR14- EventListeners, because the latter are strongly typed
- [x] Remove PageItem and use Page only. Lazy load content and segments.
- [x] Replace access via magic getters with real getters (we have it already), since the latter is faster and more explicit.
- [x] Add timestamp to css and js assets to force browsers to refresh cached css and js files
- [x] Reformat whole code base within Phpstrom according PSR-12 definitions (and defaults)
- [x] Add support for PHP 8.2
- [x] Drop support for PHP 7.4 (end-of-life)
- ~~camelCase all page properties~~ Reason: recommended coding convention for twig is snake_case
- ~~camelCase all twig filter, functions and globals including attributes~~ Reason: recommended coding convention for twig is snake_case

**Issues**
- [x] Add parent IDs to page and pageItem
- [x] Add parent route to page and pageItem
- [x] Plugin simplecontact: Success page after sending form doesn't work (index.php/contact:success), make it 2.x compatible, update readme.md
- [x] Plugin simplesearch: Make it 2.x compatible, update readme.md
- [x] Slightly refactor Event Handler and send any class to listener
- [x] Consistent handling for Twig globals as for filters, functions, and tests
- [x] Whitelist classes to be injected into plugin
- [x] Move twig "body_class" function from "twig plus" to "twig core" plugin and rename it to "css_classes"
- [x] Move twig "snippet" function from "twig_plus" to "twig_core" plugin
- [x] Move twig "page_title" function from "twig_plus" to "twig_core" plugin
- [x] Group menu twig functions to make it more clear for the user (i.e. menu_sitemap, menu_breadcrumb, etc.)
- [x] Rename twig "filter" filter due to a naming conflict with twig's own "filter" filter
- [x] Remove twig filters, functions and tests from "twig_plus" plugin that are not generic like pages_filtered() or pages_recent()
- [x] Rename "link_file" twig function to "link_media"
- [x] Rename "file" twig function to "link_file" (the above must be done first)
- [x] Remove \herbie\TwigFilter, \herbie\TwigFunction, and \herbie\TwigTest
- [x] Remove unused methods in \herbie\PageList, since we have the query method now
- [x] Rename `event` folder to `events`
- [x] Rename `herbie\sysplugin` namespace to `herbie\sysplugins`
- [x] Rename `tests` namespace to `herbie\tests`
- [x] Move commands into `commands` folder
- [x] Move middlewares into `middlewares` folder
- [x] Enhance pagination class with proper members, methods and the narrowed list of items, etc.
- [x] Use query builder in "menu_list" twig function
- [x] Implement pagination in query builder, see https://herbie.tebe.ch/doc/indepth/query-builder#paginate

**Not necessary**
- [x] ~~Update ServerRequest in Container with current Request including needed arguments like route, routeParams or page~~ Route with params can be retrieved by UrlManager

**Won't fix or unsolvable**
- robots.txt doesn't work with internal PHP server (example link http://localhost:9999/robots.txt). Reason: Seems to be an issue with the PHP web server itself.
- sitemap.xml doesn't work with internal php server (example link http://localhost:9999/sitemap.xml). Reason: Seems to be an issue with the PHP web server itself.
- Prefix all twig filters, functions and tests with `herbie_`, `h_` or similar (to prevent naming conflicts and better distinguish twig-owned filters, functions and tests). Reason: Makes filters, function or tests too clumsy.
- Evaluate and include illuminate collections in twig templates. Reason: Collections are indeed very useful, but in Twig not as practical as thought.

**Comment by tbreuss** [comment]
Possible solution for building a tree from a list. The tree could then be wrapped in a iterable class and handled with the several SPL iterator classes.

~~~php
function buildTree(array $items)
{
    $tree = [];

    /* Initialize a reference array */
    $references = [];
    foreach ($items as &$node) {

        // Add the node to our associative array using it's ID as key
        $references[$node['id']] = &$node;

        // Add empty placeholder for children
        $node['children'] = [];

        // It it's a root node, we add it directly to the tree
        if (empty($node['parent_id'])) {

            $tree[$node['id']] = &$node;

        } else {

            // It was not a root node, add this node as a reference in the parent.
            $references[$node['parent_id']]['children'][$node['id']] = &$node;

        }
    }

    return $tree;
}
~~~

**Comment by tbreuss** [comment]
Version 2.0.0 just released.
